gui/mozregui/build_runner.py

```diff
@@ -0,0 +1,187 @@
+from PyQt4.QtCore import QObject, QThread, pyqtSignal as Signal, \
+    pyqtSlot as Slot, QTimer
+
+from mozregui.global_prefs import get_prefs, apply_prefs
+from mozregression.download_manager import BuildDownloadManager
+from mozregression.test_runner import TestRunner
+from mozregression.network import get_http_session
+from mozregression.persist_limit import PersistLimit
+from mozregression.errors import LauncherError
+
+
+class GuiBuildDownloadManager(QObject, BuildDownloadManager):
+    download_progress = Signal(object, int, int)
+    download_started = Signal(object)
+    download_finished = Signal(object, str)
+
+    def __init__(self, destdir, persist_limit, **kwargs):
+        QObject.__init__(self)
+        persist_limit = PersistLimit(persist_limit)
+        BuildDownloadManager.__init__(self, destdir,
+                                      session=get_http_session(),
+                                      persist_limit=persist_limit,
+                                      **kwargs)
+
+    def _download_started(self, task):
+        self.download_started.emit(task)
+        BuildDownloadManager._download_started(self, task)
+
+    def _download_finished(self, task):
+        try:
+            self.download_finished.emit(task, task.get_dest())
+        except RuntimeError:
+            # in some cases, closing the application may destroy the
+            # underlying c++ QObject, causing this signal to fail.
+            # Skip this silently.
+            pass
+        BuildDownloadManager._download_finished(self, task)
+
+    def focus_download(self, build_info):
+        build_url, fname = self._extract_download_info(build_info)
+        dest = self.get_dest(fname)
+        build_info.build_file = dest
+        # first, stop all downloads in background (except the one for this
+        # build if any)
+        self.cancel(cancel_if=lambda dl: dest != dl.get_dest())
```

Comment:
WOW! This part of code is awesome, I haven't use lambda before even I had know it for a while!